sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/ConnectionHandler.java

```diff
@@ -0,0 +1,179 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.rabbitmq;
+
+import com.rabbitmq.client.AlreadyClosedException;
+import com.rabbitmq.client.Channel;
+import com.rabbitmq.client.Connection;
+import com.rabbitmq.client.ConnectionFactory;
+import com.rabbitmq.client.MissedHeartbeatException;
+import com.rabbitmq.client.PossibleAuthenticationFailureException;
+import com.rabbitmq.client.ShutdownSignalException;
+import java.io.Closeable;
+import java.io.IOException;
+import java.net.ProtocolException;
+import java.net.URISyntaxException;
+import java.security.KeyManagementException;
+import java.security.NoSuchAlgorithmException;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeoutException;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * RabbitMQ multiplexes over a single Connection using Channels so it should not be necessary to
+ * open multiple Connections to a single host. This class implements {@link ChannelLeaser} by
+ * maintaining a map of lessees and their associated Channels over a single Connection.
+ *
+ * <p>Note: it is unclear what, if anything, is ultimately responsible for calling close() here; it
+ * should only happen on Beam runner shutdown, or if it were knowable that there were no current
+ * Readers or Writers interacting with Rabbit.
+ */
+@ThreadSafe
+class ConnectionHandler implements ChannelLeaser, Closeable {
+  private final Map<UUID, Channel> channelsByLessee = new ConcurrentHashMap<>();
+  private final String uri;
+  private volatile Connection connection;
+
+  public ConnectionHandler(String uri) {
+    this.uri = uri;
+  }
+
+  @Override
+  public <T> T useChannel(UUID lesseeId, ChannelLeaser.UseChannelFunction<T> f) throws IOException {
+    try {
+      Channel channel = getChannel(lesseeId);
+      return f.apply(channel);
+    } catch (PossibleAuthenticationFailureException
+        | ProtocolException
+        | MissedHeartbeatException e) {
+      // full Connection-level problem
+      close();
+      throw e;
+    } catch (ShutdownSignalException e) {
+      handleShutdownSignalException(e, lesseeId);
+      throw e;
+    } catch (RuntimeException e) {
+      // may have been wrapped
+      Throwable cause = e.getCause();
+      if (cause instanceof ShutdownSignalException) {
```

Comment:
I dunno how deeply this may have been wrapped (so my guess is a single level is fine) but fyi there does exist a way to search the whole chain:
```java
Option<ShutdownSignalException> sse = Throwables.getCausalChain(e).stream().filter(t -> t instanceof ShutdownSignalException).findFirst().map(t -> (ShutdownSignalException)t);
```